### DISCUSSION TITLE
Improving developer experience by offering a stable product field API

### DISCUSSION THREAD

**Discussion by dennisnissle** [woocommerce/woocommerce#40108] [WooCommerce New Product Editor]
Hi there,

first of all: I really like the new product editing experience. One thing that IMHO is vital for its adoption and success is to offer easy ways to extend the product editor with custom fields. Both for extension developers as well as for shop owners (limited coding experience, often using small/pre-made snippets).

The issue is that the Woo Core does not offer a good API do add/register custom product fields. There are ways to embed custom product fields via Hooks, e.g. by using the `woocommerce_product_options_pricing` hook - the issue is that the Woo Core, by using this hook, does not know anything about the custom field. Instead I'd propose a real API, e.g. with custom functions like `woocommerce_register_product_field()` to register (custom) product fields. There should be some grouping/priority mechanism too e.g. to add a custom field to the pricing group.

The main advantage would be that the Woo Core could use/query these fields for many of its core functionalities, e.g. product import/export, bulk editing, REST API support which currently need additional handling for each and every custom field registered by each extension (frustrating both users, in case some of the features are lacking and developers). 

Until then, the new product editing experience (unfortunately) just adds a new abstraction layer that will frustrate users and developers, as new fields will now need to be additionally registered via a custom React logic.

I have already stated my general criticism concerning the lack of backwards compatibility within this [issue](https://github.com/woocommerce/woocommerce-blocks/issues/10733).

Best,
Dennis

**Comment by mattsherman** [comment]
@dennisnissle Thanks for sharing your thoughts!

I agree that it would be ideal to have an easy API to add custom fields to products, that would handle the following:

* Handle persisting the field to the DB
* Add REST API support
* Add field to product editor form 

Regarding that last point, the upcoming block template API handles that, and addresses (from a UI perspective), your point about:

> There should be some grouping/priority mechanism too e.g. to add a custom field to the pricing group.

Here is a an example of what that code would look like:

```php
		$group = $template->get_group( 'my_group' );

		$section = $group->add_section(
			[
				'id'         => 'example-custom-section',
				// Note: this order is relative to the section, not the template.
				'order'      => 100,
				'attributes' => [
					'title' => __( 'My custom section', 'woocommerce' ),
				],
			]
		);

		$section->add_block(
			[
				'id'         => 'example-custom-toggle',
				'order'      => 10,
				'blockName'  => 'woocommerce/product-toggle-field',
				'attributes' => [
					// Note: this property needs to be configured in the REST API.
					'property' => 'example_custom_toggle',
					'label'    => __( 'Enable extra awesomeness', 'woocommerce' ),
				],
			]
		);
```

Note that, to your point, this doesn't actually add the field to the product for the REST API (or persisting to the DB).

It would be interesting to consider whether this upcoming template API could be extended to actually handle adding the REST API support and DB persistence as well when a property was specified that wasn't already added.

Or, whether it makes more sense to have another higher-level API (like `woocommerce_register_product_field()` as you suggested) that encompassed all the things that need to be done to add a new field to a product.

> as new fields will now need to be additionally registered via a custom React logic

Note that we are striving to make it possible for developers to extend the new product editor completely through PHP in most cases, unless new UI components (blocks) are needed. We are building up a library of common block types, like checkbox, toggle, etc. that will be able to be used.

> Both for extension developers as well as for shop owners (limited coding experience, often using small/pre-made snippets).

Do you have examples of the types of custom fields that shop owners tend to add to products? I'd love to know more about what and why they are doing this, since there might be some commonalities that we could support right out of the box in the future.

**Comment by dennisnissle** [comment]
Hi @mattsherman,

thank you for your feedback! 

IMHO it would be very nice to have a consistent, high-level fields API that works for both, the current product editing experience and the new, blockified version. It would be nice if variations would be supported out-of-the-box, e.g. allow managing whether a custom field is available for variations too (override the default, parent option). 

> Do you have examples of the types of custom fields that shop owners tend to add to products?

A lot of these fields are very custom solutions depending on the type of business but here are some examples I see "in the wild":
- A shelf number, to help when picking the item from stock
- A way to select a different tax class depending on the country of the customer (part of my [OSS helper](https://wordpress.org/plugins/one-stop-shop-woocommerce/) plugin).
- File attachment options to add custom files and send them as email attachments during purchase (e.g. manual)
- Lazy-load options (AJAX) via search/autocomplete as loading all the options is simply not possible 

PS: There are so many places in Woo where (adding) fields is such an important extendibility feature, e.g products, checkout, order, order items, my-account page, registration form etc. - having a consistent fields API would create so much value both for compatibility and for the dev experience. Currently, what we have, is a mixture of hooks/filters, that all work slightly different (some handle the whole field rendering process, like in the checkout, some only work to output custom HTML - all the validation logic needs to be implemented separately). 

Best,
Dennis
